gadopt/inverse.py

```diff
@@ -359,3 +360,54 @@ def __getstate__(self):
         "Iteration Limit": 100,
     },
 }
+
+
+class DiagnosticBlock(Block):
+    def __init__(self, f, function, riesz_options={'riesz_representation': 'L2'}):
+        super().__init__()
+        self.add_dependency(function)
+        self.add_output(function.block_variable)
+        self.f = f
+        self.f_name = function.name()
+        self.riesz_options = riesz_options
+
+    def recompute_component(self, inputs, block_variable, idx, prepared):
+        return block_variable.checkpoint
+
+    def evaluate_adj_component(self, inputs, adj_inputs, block_variable, idx, prepared=None):
+        out = inputs[0]._ad_convert_type(adj_inputs[0], options=self.riesz_options)
+        out.rename('adjoint_'+self.f_name)
+        self.f.write(out)
+        return 0
+
+
+class DiagnosticConstantBlock(Block):
+    def __init__(self, constant, name):
+        super().__init__()
+        self.add_dependency(constant)
+        self.add_output(constant.block_variable)
+        self.name = name
+
+    def recompute_component(self, inputs, block_variable, idx, prepared):
+        return block_variable.checkpoint
+
+    def evaluate_adj_component(self, inputs, adj_inputs, block_variable, idx, prepared=None):
+        out = inputs[0]._ad_convert_type(adj_inputs[0])
+        print("Adjoint ", self.name, out.values()[0])
+        return 0
+
+
+class RieszL2BoundaryRepresentation:
+    """Callable that Converts l2-representatives to L2-boundary representatives"""
```

Comment:
Would be nice to have a bit more information about this...

Reply:
added a bit more based on similar comments above